zeppelin-server/src/main/java/org/apache/zeppelin/socket/ConnectionManager.java

```diff
@@ -0,0 +1,442 @@
+    }
+
+    LOGGER.debug("SEND >> " + m);
+    for (NotebookSocket conn : socketsToBroadcast) {
+      if (exclude.equals(conn)) {
+        continue;
+      }
+      try {
+        conn.send(serializeMessage(m));
+      } catch (IOException e) {
+        LOGGER.error("socket error", e);
+      }
+    }
+  }
+
+  /**
+   * Send websocket message to all connections regardless of notebook id.
+   */
+  public void broadcastToAllConnections(String serialized) {
+    broadcastToAllConnectionsExcept(null, serialized);
+  }
+
+  public void broadcastToAllConnectionsExcept(NotebookSocket exclude, String serializedMsg) {
+    synchronized (connectedSockets) {
+      for (NotebookSocket conn : connectedSockets) {
+        if (exclude != null && exclude.equals(conn)) {
+          continue;
+        }
+
+        try {
+          conn.send(serializedMsg);
+        } catch (IOException e) {
+          LOGGER.error("Cannot broadcast message to conn", e);
+        }
+      }
+    }
+  }
+
+  public Set<String> getConnectedUsers() {
+    Set<String> connectedUsers = Sets.newHashSet();
+    for (NotebookSocket notebookSocket : connectedSockets) {
+      connectedUsers.add(notebookSocket.getUser());
+    }
+    return connectedUsers;
+  }
+
+
+  public void multicastToUser(String user, Message m) {
+    if (!userSocketMap.containsKey(user)) {
+      LOGGER.warn("Multicasting to user {} that is not in connections map", user);
+      return;
+    }
+
+    for (NotebookSocket conn : userSocketMap.get(user)) {
+      unicast(m, conn);
+    }
+  }
+
+  public void unicast(Message m, NotebookSocket conn) {
+    try {
+      conn.send(serializeMessage(m));
+    } catch (IOException e) {
+      LOGGER.error("socket error", e);
+    }
+    broadcastToWatchers(StringUtils.EMPTY, StringUtils.EMPTY, m);
+  }
+
+  public void unicastParagraph(Note note, Paragraph p, String user) {
+    if (!note.isPersonalizedMode() || p == null || user == null) {
+      return;
+    }
+
+    if (!userSocketMap.containsKey(user)) {
+      LOGGER.warn("Failed to send unicast. user {} that is not in connections map", user);
+      return;
+    }
+
+    for (NotebookSocket conn : userSocketMap.get(user)) {
+      Message m = new Message(Message.OP.PARAGRAPH).put("paragraph", p);
+      unicast(m, conn);
+    }
+  }
+
+  public void broadcastNoteListExcept(List<Map<String, String>> notesInfo,
+                                      AuthenticationInfo subject) {
+    Set<String> userAndRoles;
+    NotebookAuthorization authInfo = NotebookAuthorization.getInstance();
+    for (String user : userSocketMap.keySet()) {
+      if (subject.getUser().equals(user)) {
+        continue;
+      }
+      //reloaded already above; parameter - false
+      userAndRoles = authInfo.getRoles(user);
+      userAndRoles.add(user);
+      // TODO(zjffdu) is it ok for comment the following line ?
+      // notesInfo = generateNotesInfo(false, new AuthenticationInfo(user), userAndRoles);
```

Comment:
This is the cause of https://issues.apache.org/jira/browse/ZEPPELIN-5061 and needs to be reverted